WIP ZDC Digits2raw

cortesep opened the pull request.

Commit:
Correct typos

Commit:
Adding more lineshapes

cortesep committed:
Autotrigger bits are now for all channels. Rename to Hits

cortesep committed:
Discriminator are enabled on all channels

cortesep committed:
New data member for discriminated channels that are actually used for triggering

cortesep committed:
Move AllChannelsMask to highest bit

cortesep committed:
Correcting mistake

cortesep committed:
Enable hits on non triggering channels

cortesep committed:
Introduce hits besides triggers

cortesep committed:
Allow triggering on all channels

cortesep committed:
Correct coding conventions

cortesep committed:
Updated comment

cortesep committed:
Apply clang-format

cortesep committed:
Fix link ID

cortesep committed:
Separate word trigger word for ALICE trigger

cortesep committed:
Change formatting

cortesep committed:
Change meaning of autotrigger bits, from logical (signal) to geographical (module,channel). Proper treatment of signal noise for channels that are acquired on two modules (one for trigger and readout and on another for triggering)

cortesep committed:
Removing comments

cortesep committed:
Format of GBT words

cortesep committed:
Format of GBT words

cortesep committed:
Merge branch 'dev_zdcraw' of https://github.com/cortesep/AliceO2 into dev_zdcraw

Conflicts:
	Detectors/ZDC/macro/CreateModuleConfig.C

cortesep committed:
Added structure

cortesep committed:
Small correction

cortesep committed:
New methods

cortesep committed:
Workaround compilation problem

cortesep committed:
Insert new classes

cortesep committed:
Baseline estimation. Last bunch in orbit not yet implemented

cortesep committed:
Last bunch in orbit inserted. No payload yet

cortesep committed:
Payload for last bunch

cortesep committed:
Corrected pedestals

cortesep committed:
Better formatting of output

cortesep committed:
Prepare for writing raw data

cortesep committed:
Add constant NLinks

cortesep committed:
Write raw data file

cortesep committed:
Push to dev

cortesep committed:
digi2raw executable

cortesep committed:
Correct verbosity levels

cortesep committed:
Checks on link id

cortesep committed:
Create exe

cortesep committed:
Correct name of FEEID

cortesep committed:
Apply clang-format

cortesep committed:
Apply clang-format

cortesep committed:
Apply clang-format

shahor02 commented on DataFormats/Detectors/ZDC/include/DataFormatsZDC/RawEventData.h:
```diff
@@ -0,0 +1,146 @@
+// #include <utility>
+// #include <cstring>
+
+/// \file BCData.h
```
wrong name

shahor02 commented on DataFormats/Detectors/ZDC/include/DataFormatsZDC/RawEventData.h:
```diff
@@ -0,0 +1,146 @@
+// #include "Headers/RAWDataHeader.h"
+// #include <Framework/Logger.h>
+// #include <utility>
+// #include <cstring>
```
if these includes are not needed, please remove them

shahor02 commented on DataFormats/Detectors/ZDC/include/DataFormatsZDC/RawEventData.h:
```diff
@@ -0,0 +1,146 @@
+
+constexpr unsigned short id_w0 = 0x0;
+constexpr unsigned short id_w1 = 0x1;
+constexpr unsigned short id_w2 = 0x2;
```
names of constants should start with capital

shahor02 commented on DataFormats/Detectors/ZDC/include/DataFormatsZDC/RawEventData.h:
```diff
@@ -0,0 +1,146 @@
+  unsigned empty_5 : 32;
+};
+
+struct __attribute__((__packed__)) ChannelDataV0W0 {
```
I don't see where these ``ChannelDataV0W...`` are used.

shahor02 commented on DataFormats/Detectors/ZDC/include/DataFormatsZDC/RawEventData.h:
```diff
@@ -0,0 +1,146 @@
+};
+
+struct EventData {
+  union {
+    UInt_t w[o2::zdc::NWPerBc][NWPerGBTW];
+    struct ChannelDataV0 f;
+  } data[o2::zdc::NModules][o2::zdc::NChPerModule] = {0};
```
Did you check that this initialization to 0 works? An alternative way would be to define EventData as here: https://github.com/AliceO2Group/AliceO2/blob/14851f87b3a8bc708fb0742cf9ed2d4c2f148f7a/DataFormats/Headers/include/Headers/RAWDataHeader.h#L69-L131 (with the drawback that the root cannot generate streamers for anonymous structs, but this is not needed for the raw data).

cortesep replied:
It seems to contain zeros, whatever the initializer in curly brackets. However the output structure needs to be initialized at every bunch crossing in Digits2Raw.cxx. Probably for the moment it is not needed to change the structure.

shahor02 commented on Detectors/ZDC/simulation/src/Digits2Raw.cxx:
```diff
@@ -0,0 +1,874 @@
+    return;
+  }
+
+  if (digiTree->GetBranch("ZDCDigitLabels")) {
+    digiTree->SetBranchAddress("ZDCDigitLabels", &labelsPtr);
+    LOG(INFO) << "Branch ZDCDigitLabels is connected";
+  } else {
+    LOG(INFO) << "Branch ZDCDigitLabels is missing";
+  }
```
why do you need to read lables? You can even suppress the branch to avoid root reading it automatically: ``digiTree->SetBranchStatus("ZDCDigitLabel*", 0);``

shahor02 commented on Detectors/ZDC/simulation/src/Digits2Raw.cxx:
```diff
@@ -0,0 +1,874 @@
+}
+
+//______________________________________________________________________________
+void Digits2Raw::insertLastBunch(int ibc, uint32_t orbit)
```
Cannot this method be implemented as a custom case of convertDigits (e.g. with some extra option). Most of the code is duplicated, this is quite error prone.

cortesep replied:
Good suggestion. I moved the duplicated code to private methods of the class. In this way the differences in behavior are more clear.

shahor02 commented on Detectors/ZDC/simulation/src/Digits2Raw.cxx:
```diff
@@ -0,0 +1,874 @@
+          for (Int_t iw = 0; iw < o2::zdc::NWPerBc; iw++) {
+            mFeeID = mModuleConfig->modules[im].feeID[ic];
+            std::memcpy(out, &mZDC.data[im][ic].w[iw][0], data_size);
+            mWriter.addData(mFeeID, mCruID, mLinkID, mEndPointID, ir, out_data);
```
you can actually avoid copying to out_data by using
````
gsl::span<char> payload{reinterpret_cast<char*>(&mZDC.data[im][ic].w[iw][0]), data_size};
mWriter.addData(mFeeID, mCruID, mLinkID, mEndPointID, ir, payload);
````

shahor02 commented on Detectors/ZDC/simulation/src/Digits2Raw.cxx:
```diff
@@ -0,0 +1,874 @@
+}
+
+//______________________________________________________________________________
+void Digits2Raw::emptyBunches(std::bitset<3564>& bunchPattern)
```
Is this supposed to provide empty bunches for simulated data? Note that the simulated bunch filling is stored and can be used as 
````
const auto* digctx = o2::steer::DigitizationContext::loadFromFile("collisioncontext.root");
const auto& bcfill = digctx->getBunchFilling();
for (int i = 0; i<o2::constants::lhc::LHCMaxBunches; i++) {
    if (!bf.testBC(i)) {
       printf("bc %d is not filled\n",i);
    }
````

cortesep replied:
I'm not sure of the most appropriate position when to load the filling scheme.
At present the filling scheme is extracted using this code is in the workflow executabile digi2raw.cxx
This function restricts the definition of empty bunch to remove bunches that are close to a B bunch or with beam gas from back of the calorimeter. In the future one should remove also A and C bunches (at least for the actual data taking).

shahor02 reviewed (commented):
Hi @cortesep 

Overall looks fine, please see a few comments below (the one about the RawEventData is just a suggestion, please check that the initialization to 0 is indeed done correctly).
One general comment: you are using in many places 
````
 if (cond) 
    one_line_cmd;
````
We are now trying to convert all these case to 
````
 if (cond) {
    one_line_cmd;
 }
````
Could you do this at once to avoid massive change later?

cortesep committed:
ADC limits now in Constants.h

cortesep committed:
ADC limits now in Constants.h

cortesep committed:
Correct formatting of if clauses

cortesep committed:
Remove unused code

cortesep committed:
Correct comment

cortesep committed:
Correct comment

cortesep committed:
Correct capitalization of constants

cortesep committed:
Correct clang-format

cortesep committed:
Cleanup

cortesep committed:
Move duplicated code in convertDigits and insertLastBunch to inline functions

cortesep committed:
Cleanup

cortesep committed:
Protect pointers

cortesep committed:
Correct clang-format

cortesep committed:
Avoid copy of data on output (cast pointer as suggested by Ruben) + small corrections

cortesep committed:
Read back and dump raw data

cortesep commented:
Hi Ruben,
- I tried to apply al the corrections. Probably a few if clauses are still to be corrected.
- I managed also to read back the raw data with the Detectors/ZDC/simulation/src/raw-parser.cxx
o2-raw-file-reader-workflow --input-conf ZDCraw.cfg | o2-zdc-raw-parser --log-level 2 -b > o2-zdc-raw-parser.out
The file is misplaced because I was not succesfull in compiling it in a new directory. Probably it should be moved into Detectors/ZDC/raw
- I'm not sure if this file is a good starting point to write reconstruction class. Is it safe to assume that o2-raw-file-reader-workflow will send data one TF at a time?
Thanks again,
Pietro

cortesep committed:
Correct clang-format

shahor02 commented on Detectors/ZDC/simulation/src/digi2raw.cxx:
```diff
@@ -0,0 +1,162 @@
+    add_option("verbosity,v", bpo::value<int>()->default_value(0), "verbosity level");
+    //    add_option("input-file,i", bpo::value<std::string>()->default_value(o2::base::NameConf::getDigitsFileName(o2::detectors::DetID::ZDC)),"input ZDC digits file"); // why not used?
+    add_option("input-file,i", bpo::value<std::string>()->default_value("zdcdigits.root"), "input ZDC digits file");
+    add_option("file-per-link,l", bpo::value<bool>()->default_value(false)->implicit_value(true), "create output file per CRU (default: per layer)");
```
the copy-pasted comment is misleading. We must have at least an option to write 1 file per link. As far as I can see from your code, your default is to write just 1 file. So, the help comment should be ``create output file per link (default: write single file for links)``

shahor02 commented on Detectors/ZDC/simulation/src/digi2raw.cxx:
```diff
@@ -0,0 +1,162 @@
+    add_option("output-dir,o", bpo::value<std::string>()->default_value("./"), "output directory for raw data");
+    uint32_t defRDH = o2::raw::RDHUtils::getVersion<o2::header::RAWDataHeader>();
+    add_option("rdh-version,r", bpo::value<uint32_t>()->default_value(defRDH), "RDH version to use");
+    add_option("no-empty-hbf,e", bpo::value<bool>()->default_value(false)->implicit_value(true), "do not create empty HBF pages (except for HBF starting TF)");
```
do you really need this option? This was a special feature requested by muon.

shahor02 commented on DataFormats/Detectors/ZDC/src/RawEventData.cxx:
```diff
@@ -0,0 +1,50 @@
+        // Word 2
+        printf("%04x %08x %08x ", data[im][ic].w[2][2], data[im][ic].w[2][1], data[im][ic].w[2][0]);
+        printf("%s %s %s %s %s %s 6-b ", f.Hit ? "H" : " ", f.Auto_m ? "TM" : "  ", f.Auto_0 ? "T0" : "  ", f.Auto_1 ? "T1" : "  ", f.Auto_2 ? "T2" : "  ", f.Auto_3 ? "T3" : "  ");
+        printf(" %5d %5d %5d %5d %5d %5d\n", f.s06, f.s07, f.s08, f.s09, f.s10, f.s11, f.error);
```
you have mismatch between the number of formats and number of provided args. This leads to failure of the macos CI test:
````
[ 11%] Building CXX object DataFormats/TimeFrame/CMakeFiles/O2lib-TimeFrame.dir/G__O2TimeFrame.cxx.o
/System/Volumes/Data/build/ci_checks/o2_checker_macos_o2_ci_0/sw/SOURCES/O2/4658/0/DataFormats/Detectors/ZDC/src/RawEventData.cxx:35:88: error: data argument not used by format string [-Werror,-Wformat-extra-args]
        printf(" %5d %5d %5d %5d %5d %5d\n", f.s06, f.s07, f.s08, f.s09, f.s10, f.s11, f.error);
               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                            ^
1 error generated.

````

shahor02 commented on Detectors/ZDC/simulation/include/ZDCSimulation/Digits2Raw.h:
```diff
@@ -0,0 +1,103 @@
+  EventData mZDC;                                                       /// Output structure
+  bool mIsContinuous = true;                                            /// Continuous (self-triggered) or externally-triggered readout
+  bool mOutputPerLink = false;                                          /// Split output
+  const ModuleConfig* mModuleConfig = 0;                                /// Trigger/readout configuration object
+  const SimCondition* mSimCondition = 0;                                /// Pedestal/noise configuration object
```
the new standard requires ``nullptr`` instead of 0. Due to this you get errors in https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/4658/bff9a440648e881ac50c8acce1a06c40eba16893/build_O2_fullCI/pretty.html

Please fix all of them.

shahor02 reviewed (changes requested):
Hi @cortesep 

Thanks, looks fine but please see some comments below (CI test require some changes).

There are too much commits, I would squash them to 1 commit unless you want to do more elaborate manual squashing

cortesep committed:
Corrections

cortesep commented:
Hi,
I updated the code. Now the checks are running. You can squash all the commits into one.

I should probably add some option to load the configuration objects from local files instead of downloading them from CCDB. What do you think?

Thanks,
Pietro

cortesep committed:
Corrections

cortesep committed:
Corrections

cortesep committed:
Corrections

cortesep committed:
Corrections

cortesep committed:
Corrections

shahor02 commented:
Hi @cortesep 
What is the status here? Is this PR ready for re-review?
Cheers,
 Ruben

cortesep commented:
> Hi @cortesep
> What is the status here? Is this PR ready for re-review?
> Cheers,
> Ruben

Hi Ruben,
yes. I think I implemented all the suggestions.
Ciao,
Pietro

shahor02 commented:
@cortesep sorry, missed the question:
> I should probably add some option to load the configuration objects from local files instead of downloading them from CCDB. What do you think?

I think reading from CCDB is fine, provided you can guarantee that there is a valid object there.

cortesep commented:
Hi,

no problem. In a previous message you sent me the option to use local 
CCDB do to tests and it works.


On 12/3/20 11:15 AM, Ruben Shahoyan wrote:
> @cortesep <https://github.com/cortesep> sorry, missed the question:
> 
>     I should probably add some option to load the configuration objects
>     from local files instead of downloading them from CCDB. What do you
>     think?
> 
> I think reading from CCDB is fine, provided you can guarantee that there 
> is a valid object there.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/AliceO2Group/AliceO2/pull/4658#issuecomment-737831702>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AN2A5SVH7PMEIDAZGT6JBGLSS5QMHANCNFSM4S26H7FQ>.
>

shahor02 commented:
@cortesep I was trying to test this PR, but it has conflicts with dev:
````
git checkout cortesep/dev_zdcraw
git rebase dev
First, rewinding head to replay your work on top of it...
Applying: Correct typos
Using index info to reconstruct a base tree...
M	Detectors/ZDC/macro/CreateModuleConfig.C
M	Detectors/ZDC/simulation/src/Digitizer.cxx
Falling back to patching base and 3-way merge...
Auto-merging Detectors/ZDC/macro/CreateModuleConfig.C
CONFLICT (content): Merge conflict in Detectors/ZDC/macro/CreateModuleConfig.C
error: Failed to merge in the changes.
Patch failed at 0001 Correct typos
Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".
````

Could you please resolve conflicts, rebase and forcepush the branch?

cortesep commented:
It seems the process is going forever and is quite error prone... Should I fork another branch and modify all files in one go?

cortesep commented:
Hi,

I finished the merging. It seems now I have a directory with correct 
code. Which is the following step? If I try to push it attempt to store 
everyting in the master repository.


On 12/3/20 11:22 AM, Ruben Shahoyan wrote:
> @cortesep <https://github.com/cortesep> I was trying to test this PR, 
> but it has conflicts with dev:
> 
> |git checkout cortesep/dev_zdcraw git rebase dev First, rewinding head 
> to replay your work on top of it... Applying: Correct typos Using index 
> info to reconstruct a base tree... M 
> Detectors/ZDC/macro/CreateModuleConfig.C M 
> Detectors/ZDC/simulation/src/Digitizer.cxx Falling back to patching base 
> and 3-way merge... Auto-merging Detectors/ZDC/macro/CreateModuleConfig.C 
> CONFLICT (content): Merge conflict in 
> Detectors/ZDC/macro/CreateModuleConfig.C error: Failed to merge in the 
> changes. Patch failed at 0001 Correct typos Use 'git am 
> --show-current-patch' to see the failed patch Resolve all conflicts 
> manually, mark them as resolved with "git add/rm <conflicted_files>", 
> then run "git rebase --continue". You can instead skip this commit: run 
> "git rebase --skip". To abort and get back to the state before "git 
> rebase", run "git rebase --abort". |
> 
> Could you please resolve conflicts, rebase and forcepush the branch?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/AliceO2Group/AliceO2/pull/4658#issuecomment-737841106>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AN2A5SQXFPCC2LOFPIKHDKDSS5RGJANCNFSM4S26H7FQ>.
>

shahor02 commented:
@cortesep have you done merge or rebase? You should have done the latter.
You can test if your branch cleanly merges to dev by
````
git checkout dev
git checkout -b mrgtst
git merge dev_zdcraw
````
If everything is clean, delete the mrgtst by
````
git checkout dev_zdcraw
git branch -D mrgtst
````
and fores-push your branch by
``git push -f cortesep dev_zdcraw``

cortesep commented:
Dear Ruben,
I have problems with my dev branch, however dev_zdcraw can merge into main dev branch. 
![image](https://user-images.githubusercontent.com/57937610/101030480-9f02ae00-3579-11eb-9091-335a1abca73d.png)

shahor02 commented:
What is the problem with your dev? Have you done ``git pull``?

shahor02 commented:
It seems you did not push your branch after resolving the conflicts

cortesep commented:
After resolving the conflicts it didn't compile. I probably made 
mistakes in the process or something went wrong, in summary the files 
were not the correct version.

While trying to push, however it was attempting to write to main 
repository and not to my repository.

My dev cannot automatically merge into main dev, however it seems that 
dev_zdcraw can.

Probably the cause of all the problems is that I created dev_zdcraw from 
main repository and left behind dev..

On 12/3/20 3:35 PM, Ruben Shahoyan wrote:
> It seems you did not push your branch after resolving the conflicts
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/AliceO2Group/AliceO2/pull/4658#issuecomment-738032455>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AN2A5SXJKAUFNGKLZCCJQFTSS6OZNANCNFSM4S26H7FQ>.
>

shahor02 commented:
I've rebased to dev and cleaned up your branch locally, once compiled, will test it, if everything works, I propose to close this PR and I will open it from my version. If not, I think you better fetch my version and continue from it.

cortesep commented:
Thanks Ruben. If it works, please go ahead with the merge. I was 
starting to address reconstruction but I will resume working from the 
updated dev.

On 12/3/20 5:18 PM, Ruben Shahoyan wrote:
> I've rebased to dev and cleaned up your branch locally, once compiled, 
> will test it, if everything works, I propose to close this PR and I will 
> open it from my version. If not, I think you better fetch my version and 
> continue from it.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/AliceO2Group/AliceO2/pull/4658#issuecomment-738113210>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AN2A5STIE35D3DZ6OXFKSJLSS6247ANCNFSM4S26H7FQ>.
>

shahor02 commented on Detectors/ZDC/simulation/src/Digits2Raw.cxx:
```diff
@@ -0,0 +1,704 @@
+        if (mModuleConfig->modules[im].readChannel[ic]) {
+          for (Int_t iw = 0; iw < o2::zdc::NWPerBc; iw++) {
+            gsl::span<char> payload{reinterpret_cast<char*>(&mZDC.data[im][ic].w[iw][0]), data_size};
+            mWriter.addData(mFeeID, mCruID, mLinkID, mEndPointID, ir, payload);
```
@cortesep In the test I see that only 1 link with ``feeID=0xf`` gets some data and in the code I don't see where do you set this ``mFeeID``: it remains on 0xf because this was the last registered link.
I was testing as:
````
2-sim -m ZDC PIPE -n 5 -g pythia8hi  --configKeyValues "Diamond.width[2]=6.; SimCutParams.maxRTracking=100;"
o2-sim-digitizer-workflow  --interactionRate 10000
o2-zdc-digi2raw -o raw/ZDC --file-per-link 
o2-raw-file-check --input-conf raw/ZDC/ZDCraw.cfg -v2 | grep Offs | grep -v 'OffsNext:64'
[INFO] EP:0 CRU:0x0000 Link:0   FEEID:0x000f SrcID:ZDC[15] Packet:4   MemSize:4960  OffsNext:4960   prio.:0 HS:64 HV:6
[INFO] EP:0 CRU:0x0000 Link:0   FEEID:0x000f SrcID:ZDC[15] Packet:6   MemSize:1312  OffsNext:1312   prio.:0 HS:64 HV:6
[INFO] EP:0 CRU:0x0000 Link:0   FEEID:0x000f SrcID:ZDC[15] Packet:8   MemSize:2464  OffsNext:2464   prio.:0 HS:64 HV:6
[INFO] EP:0 CRU:0x0000 Link:0   FEEID:0x000f SrcID:ZDC[15] Packet:10  MemSize:3136  OffsNext:3136   prio.:0 HS:64 HV:6
[INFO] EP:0 CRU:0x0000 Link:0   FEEID:0x000f SrcID:ZDC[15] Packet:12  MemSize:2560  OffsNext:2560   prio.:0 HS:64 HV:6

````

shahor02 commented:
BTW, as the next priority (after fixing this PR), I would consider raw->digits decoder, do you already have it?

cortesep commented:
Hi,

I don't have it. However it will be faster to implement w.r.t. 
reconstruction. I wonder if the reconstruction should use digits or raw 
data. At present the digits don't have all necessary information to 
reconstruct and that's why in digi2reco there is some code to add the 
pedestal information.

On 12/3/20 9:19 PM, Ruben Shahoyan wrote:
> BTW, as the next priority (after fixing this PR), I would consider 
> raw->digits decoder, do you already have it?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/AliceO2Group/AliceO2/pull/4658#issuecomment-738284054>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AN2A5SUJTYW6LV6FGHGYXSDSS7XGHANCNFSM4S26H7FQ>.
>

cortesep commented:
Hi Ruben,

I should correct this file and assign correct mLinkID.
	    mFeeID=2*im+ic/2;

There are also many more problems. It seems that the branch dev_zdcraw 
is not in sync with my local directory. Many files are missing even if I 
did the commits. I don't understand what happened. For example the 
directory Detectors/ZDC/raw does not exists
and the file Detectors/ZDC/simulation/src/raw-parser.cxx
should be ./Detectors/ZDC/raw/src/raw-parser.cxx

A possibility would be to delete

Detectors/ZDC/simulation/src/raw-parser.cxx

and then I add the missing pieces after the merge. Or I start to work on 
your corrected version and then merge.

Pietro

On 12/3/20 9:17 PM, Ruben Shahoyan wrote:
> *@shahor02* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In Detectors/ZDC/simulation/src/Digits2Raw.cxx 
> <https://github.com/AliceO2Group/AliceO2/pull/4658#discussion_r535559180>:
> 
>> +    bool T2 = mZDC.data[im][0].f.Auto_2;
> +    bool T3 = mZDC.data[im][0].f.Auto_3;
> +    bool A0 = mZDC.data[im][0].f.Alice_0;
> +    bool A1 = mZDC.data[im][0].f.Alice_1;
> +    bool A2 = mZDC.data[im][0].f.Alice_2;
> +    bool A3 = mZDC.data[im][0].f.Alice_3;
> +    bool tcond_continuous = T0 || T1;
> +    bool tcond_triggered = A0 || A1 || (A2 && (T0 || TM)) || (A3 && T0);
> +    bool tcond_last = mZDC.data[im][0].f.bc == 3563;
> +    // Condition to write GBT data
> +    if (tcond_triggered || (mIsContinuous && tcond_continuous) || (mZDC.data[im][0].f.bc == 3563)) {
> +      for (UInt_t ic = 0; ic < o2::zdc::NChPerModule; ic++) {
> +        if (mModuleConfig->modules[im].readChannel[ic]) {
> +          for (Int_t iw = 0; iw < o2::zdc::NWPerBc; iw++) {
> +            gsl::span<char> payload{reinterpret_cast<char*>(&mZDC.data[im][ic].w[iw][0]), data_size};
> +            mWriter.addData(mFeeID, mCruID, mLinkID, mEndPointID, ir, payload);
> 
> @cortesep <https://github.com/cortesep> In the test I see that only 1 
> link with |feeID=0xf| gets some data and in the code I don't see where 
> do you set this |mFeeID|: it remains on 0xf because this was the last 
> registered link.
> I was testing as:
> 
> |2-sim -m ZDC PIPE -n 5 -g pythia8hi --configKeyValues 
> "Diamond.width[2]=6.; SimCutParams.maxRTracking=100;" 
> o2-sim-digitizer-workflow --interactionRate 10000 o2-zdc-digi2raw -o 
> raw/ZDC --file-per-link o2-raw-file-check --input-conf 
> raw/ZDC/ZDCraw.cfg -v2 | grep Offs | grep -v 'OffsNext:64' [INFO] EP:0 
> CRU:0x0000 Link:0 FEEID:0x000f SrcID:ZDC[15] Packet:4 MemSize:4960 
> OffsNext:4960 prio.:0 HS:64 HV:6 [INFO] EP:0 CRU:0x0000 Link:0 
> FEEID:0x000f SrcID:ZDC[15] Packet:6 MemSize:1312 OffsNext:1312 prio.:0 
> HS:64 HV:6 [INFO] EP:0 CRU:0x0000 Link:0 FEEID:0x000f SrcID:ZDC[15] 
> Packet:8 MemSize:2464 OffsNext:2464 prio.:0 HS:64 HV:6 [INFO] EP:0 
> CRU:0x0000 Link:0 FEEID:0x000f SrcID:ZDC[15] Packet:10 MemSize:3136 
> OffsNext:3136 prio.:0 HS:64 HV:6 [INFO] EP:0 CRU:0x0000 Link:0 
> FEEID:0x000f SrcID:ZDC[15] Packet:12 MemSize:2560 OffsNext:2560 prio.:0 
> HS:64 HV:6 |
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/AliceO2Group/AliceO2/pull/4658#pullrequestreview-544381457>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AN2A5STCQGY5V5U5N354XXLSS7W4VANCNFSM4S26H7FQ>.
>

shahor02 commented:
Hi @cortesep ,
For the reconstruction: usually, detectors implement raw->digits decoder as a separate device then the reconstruction device works with digits input, the only difference between the real and MC data being that in the latter case the reconstruction also takes/accounts the MC labels input.

For the lost directories: I propose that I close this PR, push/merge my cleaned up version (it is already rebase to dev), then you continue development from dev. To move the files you should use git commands, not unix, e.g. 
````
cd O2/Detectors/ZDC
mkdir -p raw/src
git mv reconstruction/src/raw-parser.cxx raw/src/
````

Cheers,
 Ruben
